### PR TITLE
SC-1109 ReadUserById and Search User APIs E to E implementation

### PR DESCRIPTION
**Pull request by anmol2302**
<h2>ReadUserById and Search User APIs E to E implementation</h2>
<b>Ticket Ref: </b>[SC-1109](https://project-sunbird.atlassian.net/browse/SC-1109) </br>
1: ReadUserById & Search user APIs implemeted </br>
2: ResponseCode added for Bad_Request </br>
3: Method for handle the success response and failure response added in RequestHandler </br>
4: ElasticSearch config changes (timeout and promise handling)

### CODE REVIEW

File: user-org-actor/src/main/java/org/sunbird/Application.java [hunk 1]
```diff
@@ -22,11 +22,11 @@
...
-    private Application() { }
+    private Application() {
+    }
 
     // static method to create instance of ActorService class
-    public static Application getInstance()
-    {
+    public static Application getInstance() {
         if (instance == null)
             instance = new Application();
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/Application.java]: Remove line 21 :
private static Application instance =  new Application();
and  getInstance() method will return instance;

File: user-org-actor/src/main/java/org/sunbird/Application.java [hunk 2]
```diff
@@ -37,16 +37,11 @@ public static Application getInstance()
     public void init() {
         List<String> actorClassPaths = new ArrayList<>();
         actorClassPaths.add("org.sunbird");
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/Application.java:34]: "org.sunbird" is this base package name. take it as constant and add doc on that constant ,

File: user-org-actor/src/main/java/org/sunbird/Application.java [hunk 3]
```diff
@@ -37,16 +37,11 @@ public static Application getInstance()
...
-        ActorService.getInstance().init("userOrgActorSystem",actorClassPaths);
+        ActorService.getInstance().init("userOrgActorSystem", actorClassPaths);
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/Application.java]: userOrgActorSystem can be any name or this is fixed name.
Use Constant here.

File: user-org-actor/src/main/java/org/sunbird/SearchDtoMapper.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
+package org.sunbird;
+
+import org.sunbird.dto.SearchDTO;
+
+import java.math.BigInteger;
+import java.util.*;
+
+public class SearchDtoMapper {
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/SearchDtoMapper.java]: java doc missing.

File: user-org-actor/src/main/java/org/sunbird/SearchDtoMapper.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+
+    public static SearchDTO createSearchDto(Map<String, Object> searchQueryMap) {
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/SearchDtoMapper.java]: Method doc missing.

File: user-org-actor/src/main/java/org/sunbird/SearchDtoMapper.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        SearchDTO search = new SearchDTO();
+        if (searchQueryMap.containsKey("query")) {
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/SearchDtoMapper.java]: Each if can be part of private method.

File: user-org-actor/src/main/java/org/sunbird/user/UserReadActor.java [hunk 7]
```diff
@@ -17,16 +24,24 @@
 )
 public class UserReadActor extends BaseActor {
 
+    static ElasticSearchService es = EsClientFactory.getInstance("rest");
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserReadActor.java]: use EsClientFactory .EsClient

File: user-org-actor/src/main/java/org/sunbird/user/UserReadActor.java [hunk 8]
```diff
@@ -17,16 +24,24 @@
...
+
     @Override
     public void onReceive(Request request) throws Throwable {
-        if(request.getOperation().equalsIgnoreCase("readUserById")){
-            Response response = new Response();
-            response.put("userId", "123-456-7890");
-            response.put("firstName", "Amit");
-            response.put("lastName", "kumar");
-            sender().tell(response,self());
+        if (request.getOperation().equalsIgnoreCase("readUserById")) {
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserReadActor.java]: readUserById need to be read from Constant.

File: user-org-actor/src/main/java/org/sunbird/user/UserReadActor.java [hunk 9]
```diff
@@ -17,16 +24,24 @@
...
+            readUserById(request);
         } else {
             onReceiveUnsupportedMessage("UserCreateActor");
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserReadActor.java]: use Constant to read UserCreateActor

File: user-org-actor/src/main/java/org/sunbird/user/UserReadActor.java [hunk 10]
```diff
@@ -17,16 +24,24 @@
...
         }
     }
+
+    public void readUserById(Request request) {
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserReadActor.java]: Method Doc missing.

File: user-org-actor/src/main/java/org/sunbird/user/UserReadActor.java [hunk 11]
```diff
@@ -17,16 +24,24 @@
...
+        String userId = (String) request.getRequest().get("userId");
+        Future<Map<String, Object>> future = es.getDataByIdentifier("user", userId);
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserReadActor.java]: This code need to be moved in DAO layer.
no String should be hard coded.

File: user-org-actor/src/main/java/org/sunbird/user/UserSearchActor.java [hunk 12]
```diff
@@ -18,17 +27,24 @@
 )
 public class UserSearchActor extends BaseActor {
 
+    static ElasticSearchService es= EsClientFactory.getInstance("rest");
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserSearchActor.java]: Use Predefine enum to read client.

File: user-org-actor/src/main/java/org/sunbird/user/UserSearchActor.java [hunk 13]
```diff
@@ -18,17 +27,24 @@
...
+
     @Override
     public void onReceive(Request request) throws Throwable {
         if(request.getOperation().equalsIgnoreCase("searchUser")){
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserSearchActor.java]: searchUser need to be Constant.

Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserSearchActor.java]: request.getOperation().equalsIgnoreCase("searchUser") need to be reverse check to avoid NPE.

File: user-org-actor/src/main/java/org/sunbird/user/UserSearchActor.java [hunk 14]
```diff
@@ -18,17 +27,24 @@
...
-            response.put("lastName", "kumar");
-            sender().tell(response,self());
+            searchUser(request);
         } else {
-            onReceiveUnsupportedMessage("UserCreateActor");
+            onReceiveUnsupportedMessage("UserSearchActor");
         }
     }
 
+    public void  searchUser(Request request) {
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserSearchActor.java]: Method doc missing.

File: user-org-actor/src/main/java/org/sunbird/user/UserSearchActor.java [hunk 15]
```diff
@@ -18,17 +27,24 @@
...
+        SearchDTO searchDTO= SearchDtoMapper.createSearchDto(request.getRequest());
+        Future<Map<String, Object>> future = es.search(searchDTO, "user");
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserSearchActor.java]: user used Constant.

File: user-org-actor/src/main/java/org/sunbird/user/UserSearchActor.java [hunk 16]
```diff
@@ -18,17 +27,24 @@
...
+        Map<String, Object> responseMap = (Map) ElasticSearchHelper.getResponseFromFuture(future);
+        Response response=new Response();
+        response.getResult().put("response",responseMap);
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserSearchActor.java]: use Constant.

File: setup.md [hunk 17]
```diff
@@ -1,3 +1,5 @@
 ## Pre-requisites
 # Environment Variables
 
+sunbird_m2_repo_path=file://~/.m2/repository
```
Comment (manzarul) [inline setup.md]: what the use of this : sunbird_m2_repo_path

File: user-org-actor/src/main/java/org/sunbird/user/dao/UserDaoFactory.java [hunk 18]
```diff
@@ -20,13 +20,20 @@ public static UserDaoFactory getInstance()
...
 
-    public IUserDao getDaoImpl(String daoImplType){
+
+    private IUserOSDao userDao;
+
+    public Object getDaoImpl(String daoImplType){
 
         if(daoImplType.equalsIgnoreCase(DaoImplType.OPEN_SABER.getType())) {
-            return new UserOpenSaberDaoImpl();
+            return new UserOpenSaberOSDaoImpl();
```
Comment (indrajra) [inline user-org-actor/src/main/java/org/sunbird/user/dao/UserDaoFactory.java]: Isnt' it odd to read? OpenSABER and then OS again...please use one. Similarly for the false part of this clause too - ES, OS. Keep one.

File: user-org-actor/src/main/java/org/sunbird/user/dao/UserEsOSDaoImpl.java [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+    public Response searchUser(Map<String, Object> request) {
+        SearchDTO searchDTO = SearchDtoMapper.createSearchDto(request);
+        Future<Map<String, Object>> future = es.search(searchDTO, EsIndex.USER.getIndexName());
+        Map<String, Object> responseMap = (Map) ElasticSearchHelper.getResponseFromFuture(future);
+        return prepareEsSuccessResponse(responseMap);
+    }
+
+
+    private Response prepareEsSuccessResponse(Map<String, Object> responseMap) {
+        response = new Response();
```
Comment (indrajra) [inline user-org-actor/src/main/java/org/sunbird/user/dao/UserEsOSDaoImpl.java]: If this is the only method to return Response, why declare this as field. Create local var.

File: user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+        search = setSearchLimitInRequest(search);
+        search = setSearchLimitWhenOffsetIsPresentInRequest(search);
+        return search;
+    }
+
+
+    public static SearchDTO searchForQueryInRequest(Map<String, Object> searchQueryMap, SearchDTO search) {
+        if (searchQueryMap.containsKey("query")) {
+            search.setQuery((String) searchQueryMap.get("query"));
+            return search;
```
Comment (indrajra) [inline user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java]: remove return.

File: user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+
+public class SearchDtoMapper {
+
+
+    /**
+     * this method will create SearchDto pojo object.
+     * @param searchQueryMap
+     * @return
+     */
+    public static SearchDTO createSearchDto(Map<String, Object> searchQueryMap) {
```
Comment (indrajra) [inline user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java]: Dislike static methods...Create a factory/mapper class that takes in Map and return a SearchDTO.

File: user-org-actor/src/main/java/org/sunbird/Application.java [hunk 22]
```diff
@@ -5,42 +5,43 @@
...
     }
 
     // instantiate actor system and actors
     public void init() {
         List<String> actorClassPaths = new ArrayList<>();
         actorClassPaths.add("org.sunbird");
-        ActorService.getInstance().init("userOrgActorSystem",actorClassPaths);
-
+        ActorService.getInstance().init(USER_ORG_ACTOR_SYSTEM,actorClassPaths);
         OpenSaberApplication.main(new String[0]);
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/Application.java:36]: We don't need to call main method. 
When we call OpenSaberApplication.getContext(); Open saber will internally call main method.

File: user-org-actor/src/main/java/org/sunbird/user/UserCreateActor.java [hunk 23]
```diff
@@ -24,16 +24,16 @@
     private Response response = null;
 
     @Override
-    public void onReceive(Request  request) throws Throwable {
-        if(request.getOperation().equalsIgnoreCase(UserActorOperations.CREATE_USER.getOperation())){
+    public void onReceive(Request request) {
+        if (UserActorOperations.CREATE_USER.getOperation().equalsIgnoreCase(request.getOperation())) {
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserCreateActor.java:33]: Add logger here.

File: user-org-actor/src/main/java/org/sunbird/user/UserCreateActor.java [hunk 24]
```diff
@@ -24,16 +24,16 @@
...
             createUser(request);
         } else {
-            onReceiveUnsupportedMessage("UserCreateActor");
+            onReceiveUnsupportedMessage(this.getClass().getName());
         }
     }
 
     private void createUser(Request request) {
-        IUserDao userDao = UserDaoFactory.getInstance().getDaoImpl(DaoImplType.OPEN_SABER.getType());
+        IUserOSDao userDao = (IUserOSDao) UserDaoFactory.getInstance().getDaoImpl(DaoImplType.OPEN_SABER.getType());
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserCreateActor.java]: Add some logger

File: user-org-actor/src/main/java/org/sunbird/user/UserCreateActor.java [hunk 25]
```diff
@@ -24,16 +24,16 @@
...
         try {
             response = userDao.addUser(request.getRequest());
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserCreateActor.java]: follow same naming either createUser or addUser not both.

File: user-org-actor/src/main/java/org/sunbird/user/UserCreateActor.java (hunk 24 above)
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserCreateActor.java]: Why not getInstance itself return me DAO impl instance based on passed enum.

File: user-org-actor/src/main/java/org/sunbird/user/UserReadActor.java [hunk 26]
```diff
@@ -17,16 +24,21 @@
 )
 public class UserReadActor extends BaseActor {
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserReadActor.java:29]: Add java code for class

File: user-org-actor/src/main/java/org/sunbird/user/UserReadActor.java [hunk 27]
```diff
@@ -17,16 +24,21 @@
...
+
     @Override
     public void onReceive(Request request) throws Throwable {
-        if(request.getOperation().equalsIgnoreCase("readUserById")){
-            Response response = new Response();
-            response.put("userId", "123-456-7890");
-            response.put("firstName", "Amit");
-            response.put("lastName", "kumar");
-            sender().tell(response,self());
+        if (UserActorOperations.READ_USER_BY_ID.getOperation().equalsIgnoreCase(request.getOperation())) {
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserReadActor.java:36]: Add logger that will help in debugging issues.

File: user-org-actor/src/main/java/org/sunbird/user/UserReadActor.java [hunk 28]
```diff
@@ -17,16 +24,21 @@
...
+            readUserById(request);
         } else {
-            onReceiveUnsupportedMessage("UserCreateActor");
+            onReceiveUnsupportedMessage(this.getClass().getName());
         }
     }
+
+    public void readUserById(Request request) throws BaseException {
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserReadActor.java:49]: Java doc for method.

File: user-org-actor/src/main/java/org/sunbird/user/UserSearchActor.java [hunk 29]
```diff
@@ -18,17 +22,22 @@
 )
 public class UserSearchActor extends BaseActor {
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserSearchActor.java:29]: Java doc for class missing.

File: user-org-actor/src/main/java/org/sunbird/user/UserSearchActor.java [hunk 30]
```diff
@@ -18,17 +22,22 @@
...
+
     @Override
     public void onReceive(Request request) throws Throwable {
-        if(request.getOperation().equalsIgnoreCase("searchUser")){
-            Response response = new Response();
-            response.put("userId", "123-456-7890");
-            response.put("firstName", "Amit");
-            response.put("lastName", "kumar");
-            sender().tell(response,self());
+
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/UserSearchActor.java:36]: logs need to be added.

File: user-org-actor/src/main/java/org/sunbird/user/dao/IUserESDao.java [hunk 31]
```diff
@@ -0,0 +1,29 @@
...
+
+import org.sunbird.exception.BaseException;
+import org.sunbird.response.Response;
+
+import java.util.Map;
+
+/**
+ * @author anmolgupta
+ */
+public interface IUserESDao {
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/dao/IUserESDao.java:11]: Add java doc

File: user-org-actor/src/main/java/org/sunbird/user/dao/IUserESDao.java [hunk 32]
```diff
@@ -0,0 +1,29 @@
...
+
+    /**
+     * this method will be used to get user by id from elastic search.
+     * @param userId
+     * @return
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/dao/IUserESDao.java]: why return is empty.

File: user-org-actor/src/main/java/org/sunbird/user/dao/IUserESDao.java [hunk 33]
```diff
@@ -0,0 +1,29 @@
...
+    public Response getUserById(String userId) throws BaseException;
+
+
+    /**
+     * this method will be used to search user from elasticsearch
+     * @param request
+     * @return
+     * @throws BaseException
+     */
+    public Response searchUser(Map<String, Object> request) throws BaseException;
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/dao/IUserESDao.java]: why not this method will take SearchDto instead of map

File: user-org-actor/src/main/java/org/sunbird/user/dao/IUserOSDao.java [hunk 34]
```diff
@@ -11,8 +11,10 @@
 /**
  * @author Amit Kumar
  */
-public interface IUserDao {
+public interface IUserOSDao {
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/dao/IUserOSDao.java:14]: Java doc for class is misisng

File: user-org-actor/src/main/java/org/sunbird/user/dao/UserDaoFactory.java (hunk 18 above)
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/dao/UserDaoFactory.java]: Why not it's singlton?

File: user-org-actor/src/main/java/org/sunbird/user/dao/UserEsOSDaoImpl.java [hunk 35]
```diff
@@ -0,0 +1,58 @@
...
+import org.sunbird.util.jsonkey.JsonKey;
+import scala.concurrent.Future;
+
+import java.util.Map;
+
+
+/**
+ * @author anmolgupta
+ */
+public class UserEsOSDaoImpl implements IUserESDao {
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/dao/UserEsOSDaoImpl.java]: add java doc for class

File: user-org-actor/src/main/java/org/sunbird/user/dao/UserEsOSDaoImpl.java [hunk 36]
```diff
@@ -0,0 +1,58 @@
...
+
+    private ElasticSearchService es = EsClientFactory.getInstance(EsClientFactory.EsClient.REST.getName());
+    private Localizer localizer = Localizer.getInstance();
+    private Response response = null;
+
+    @Override
+    public Response getUserById(String userId) throws BaseException {
+        Future<Map<String, Object>> future = es.getDataByIdentifier(EsIndex.USER.getIndexName(), userId);
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/dao/UserEsOSDaoImpl.java]: Please add logger , without logger it's would be very difficult to find issue on server.

File: user-org-actor/src/main/java/org/sunbird/user/dao/UserEsOSDaoImpl.java [hunk 37]
```diff
@@ -0,0 +1,58 @@
...
+        if (MapUtils.isNotEmpty(responseMap)) {
+            return prepareEsSuccessResponse(responseMap);
+        } else {
+            throw new UserException.UserNotFound(IResponseMessage.USER_NOT_FOUND, localizer.getMessage(IResponseMessage.USER_NOT_FOUND, null), ResponseCode.BAD_REQUEST.getCode());
+        }
+    }
+
+    @Override
+    public Response searchUser(Map<String, Object> request) {
+        SearchDTO searchDTO = SearchDtoMapper.createSearchDto(request);
```
Comment (manzarul) [inline user-org-actor/src/main/java/org/sunbird/user/dao/UserEsOSDaoImpl.java]: Search Dto need to be supply to that method.

File: user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java [hunk 38]
```diff
@@ -0,0 +1,166 @@
...
+        SearchDTO search = new SearchDTO();
+        search = searchForQueryInRequest(searchQueryMap, search);
```
Comment (indrajra) [inline user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java]: If you create a separate class, there is no need to pass the second param (search)...You can simply do a series of setters and finally say, build() to return you a SearchDTO instance.

Also, rename searchForFieldsInRequest -> setFields()
searchForSortByInRequest -> setSortBy() and alike. Keep names simple, they are setting, not searching...

File: user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java [hunk 39]
```diff
@@ -0,0 +1,166 @@
...
+import java.util.*;
+
+/**
+ * This class will be used to map the SearchDto
+ *
+ * @author anmolgupta
+ */
+
+
+public class SearchDtoMapper {
```
Comment (manzarul) [inline user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java:13]: Java doc for class misisng.

File: user-org-es-util/src/main/java/org/sunbird/es/service/impl/ElasticSearchRestHighImpl.java [hunk 40]
```diff
@@ -212,6 +212,7 @@ public void onResponse(GetResponse getResponse) {
                 promise.success(new HashMap<>());
```
Comment (indrajra) [inline user-org-es-util/src/main/java/org/sunbird/es/service/impl/ElasticSearchRestHighImpl.java]: init hashmap with empty...and then have one return path. 3 promise.success and if-else conditions isn't good to read. Since there is no failure, we can init with empty.

File: user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java [hunk 41]
```diff
@@ -0,0 +1,166 @@
...
+
+
+    /**
+     * this method will create SearchDto pojo object.
+     * @param searchQueryMap
+     * @return
```
Comment (manzarul) [inline user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java:36]: update return object.

File: user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java (hunk 38 above)
Comment (manzarul) [inline user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java]: Same search object is getting modified every time , so we can avaoid reassign 
search = searchForQueryFieldsInRequest(searchQueryMap, search);
 can be  searchForQueryFieldsInRequest(searchQueryMap, search); only

File: user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java [hunk 42]
```diff
@@ -0,0 +1,166 @@
...
+        search = searchForOffsetInRequest(searchQueryMap, search);
+        search = searchForSoftConstraintsInRequest(searchQueryMap, search);
+        search = searchForGroupQueryInRequest(searchQueryMap, search);
+        search = setSearchLimitInRequest(search);
+        search = setSearchLimitWhenOffsetIsPresentInRequest(search);
+        return search;
+    }
+
+
+    public static SearchDTO searchForQueryInRequest(Map<String, Object> searchQueryMap, SearchDTO search) {
```
Comment (manzarul) [inline user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java]: add method java doc.
any public method must have java doc.

File: user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java [hunk 43]
```diff
@@ -0,0 +1,166 @@
...
+        if (searchQueryMap.containsKey("query")) {
```
Comment (manzarul) [inline user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java]: All hardcode string need to be put inside constant.

File: user-org-service/app/controllers/usermanagement/UserController.java [hunk 44]
```diff
@@ -21,7 +21,7 @@
    */
   public CompletionStage<Result> createUser() {
     startTrace("createUser");
-    CompletionStage<Result> response = createHandelRequest(request());
+    CompletionStage<Result> response = handleRequest(request(),null,"createUser");
```
Comment (indrajra) [inline user-org-service/app/controllers/usermanagement/UserController.java]: Can we think of the operation be set, rather than pass? BaseController can expose a protected operationName field, which the derived classes can set. Also, if there is httpExecutionContext going to be null, please create an overload of handleRequest - rather than passing null.

File: user-org-es-util/src/main/java/org/sunbird/helper/EsIndex.java [hunk 45]
```diff
@@ -0,0 +1,16 @@
+package org.sunbird.helper;
+
+public enum EsIndex {
```
Comment (manzarul) [inline user-org-es-util/src/main/java/org/sunbird/helper/EsIndex.java:7]: add java doc

File: user-org-util/pom.xml [hunk 46]
```diff
@@ -49,5 +49,17 @@
             <version>${junit.version}</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
```
Comment (indrajra) [inline user-org-util/pom.xml:52]: think this (everit schema) was unrelated change to this PR. remove it from this PR.

**Review by indrajra — CHANGES_REQUESTED** [review]

File: user-org-service/app/controllers/BaseController.java [hunk 47]
```diff
@@ -177,9 +170,9 @@ public String jsonifyResponseObject(Response response) {
         Response response = new Response();
         Http.RequestBody requestBody = request().body();
```
Comment (manzarul) [inline user-org-service/app/controllers/BaseController.java]: What will happen when         
Http.RequestBody requestBody = request().body();  is null means we have get request
and add some logger here

File: user-org-service/app/controllers/BaseController.java [hunk 48]
```diff
@@ -177,9 +170,9 @@ public String jsonifyResponseObject(Response response) {
...
         Request request = null;
-        try{
+        try {
             request = (Request) RequestMapper.mapRequest(requestBody.asJson(), Request.class);
-        } catch (Exception ex){
+        } catch (Exception ex) {
             ex.printStackTrace();
```
Comment (manzarul) [inline user-org-service/app/controllers/BaseController.java]: all exception should have logger

File: user-org-service/app/controllers/BaseController.java [hunk 49]
```diff
@@ -203,6 +196,23 @@ public String jsonifyResponseObject(Response response) {
...
+    }
+
+
+    /**
+     * this method will map the play request to the org.sunbird.request.Request class.
+     * @param requestBody
+     * @return
+     */
+
+    public static Request mapRequest(Http.Request requestBody) {
```
Comment (manzarul) [inline user-org-service/app/controllers/BaseController.java]: return is missing.

File: user-org-service/app/controllers/BaseController.java [hunk 50]
```diff
@@ -203,6 +196,23 @@ public String jsonifyResponseObject(Response response) {
...
+
+        Request request = new Request();
+        try {
+            return (Request) RequestMapper.mapRequest(requestBody.body().asJson(), Request.class);
+        } catch (Exception e) {
+            return request;
```
Comment (manzarul) [inline user-org-service/app/controllers/BaseController.java]: add logger here. 
and why we are returning empty Request object. How will u handle if request is incorrect format

File: user-org-service/app/controllers/RequestHandler.java [hunk 51]
```diff
@@ -2,94 +2,100 @@
...
 import scala.concurrent.Await;
 import scala.concurrent.Future;
-import utils.RequestMapper;
 
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
 import java.util.concurrent.TimeUnit;
 
-public class RequestHandler  extends BaseController{
+public class RequestHandler extends BaseController {
```
Comment (manzarul) [inline user-org-service/app/controllers/RequestHandler.java:30]: Add java doc.

File: user-org-service/app/controllers/RequestHandler.java [hunk 52]
```diff
@@ -2,94 +2,100 @@
...
-                Response res = (Response)obj;
-                cf.complete(String.valueOf(Json.toJson(res)));
-            }
-        } catch (Exception e) {
-            e.printStackTrace();
-        }
-        endTrace("handelRequest");
-        return cf.thenApplyAsync(Results::ok, httpExecutionContext.current());
+        request.setOperation(operation);
+        ProjectLogger.log("UserController:createUser :: Requested operation " + request.getOperation());
```
Comment (manzarul) [inline user-org-service/app/controllers/RequestHandler.java]: log level is missing here.

File: user-org-service/app/controllers/usermanagement/UserController.java (hunk 44 above)
Comment (anmol2302) [reply user-org-service/app/controllers/usermanagement/UserController.java]: in place of null, we will pass validation function (lambda expression) right now since validation is not there in this pr so I have place null over here.

File: user-org-actor/src/main/java/org/sunbird/BaseActor.java [hunk 53]
```diff
@@ -25,9 +25,11 @@ public void onReceive(Object message) throws Throwable {
         if (message instanceof Request) {
             Request request = (Request) message;
             String operation = request.getOperation();
-            ProjectLogger.log("BaseActor: onReceive called for operation: " + operation, LoggerEnum.INFO);
+            ProjectLogger.log("BaseActor:onReceive called for operation:" + operation, LoggerEnum.INFO);
             try {
+                ProjectLogger.log(String.format("%s:%s:method started at %s",this.getClass().getSimpleName(),operation,System.currentTimeMillis()), LoggerEnum.DEBUG);
                 onReceive(request);
+                ProjectLogger.log(String.format("%s:%s:method started at %s",this.getClass().getSimpleName(),operation,System.currentTimeMillis()), LoggerEnum.DEBUG);
```
Comment (indrajra) [inline user-org-actor/src/main/java/org/sunbird/BaseActor.java]: started at ? - why dont we subtract and add another total time field in the log. Otherwise, this work is getting to the reader. 
Because we use ProjectLogger.log, we have to pass these getClass. I've not seen any projects that use a common Logger like we do :-(. I find we must remove it completely.

File: user-org-actor/src/main/java/org/sunbird/user/UserCreateActor.java [hunk 54]
```diff
@@ -32,10 +36,16 @@ public void onReceive(Request request) {
         }
     }
 
+    /**
+     * this method is used to create the user
+     * @param request
+     */
     private void createUser(Request request) {
-        IUserOSDao userDao = (IUserOSDao) UserDaoFactory.getInstance().getDaoImpl(DaoImplType.OPEN_SABER.getType());
+        ProjectLogger.log(String.format("%s:%s:method started at %s",this.getClass().getSimpleName(),"createUser",System.currentTimeMillis()), LoggerEnum.DEBUG.name());
```
Comment (indrajra) [inline user-org-actor/src/main/java/org/sunbird/user/UserCreateActor.java]: Why are you not using startTrace and endTrace you introduced in another previous PR?

File: user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java [hunk 55]
```diff
@@ -15,126 +15,131 @@
...
-        search = searchForExistsInRequest(searchQueryMap, search);
-        search = searchForNotExistsInRequest(searchQueryMap, search);
-        search = searchForSortByInRequest(searchQueryMap, search);
-        search = searchForLimitInRequest(searchQueryMap, search);
-        search = searchForOffsetInRequest(searchQueryMap, search);
-        search = searchForSoftConstraintsInRequest(searchQueryMap, search);
-        search = searchForGroupQueryInRequest(searchQueryMap, search);
-        search = setSearchLimitInRequest(search);
-        search = setSearchLimitWhenOffsetIsPresentInRequest(search);
+        setQuery(searchQueryMap, search);
```
Comment (indrajra) [inline user-org-es-util/src/main/java/org/sunbird/dto/SearchDtoMapper.java]: This still looks odd, sorry. Better in function names, but not in functional richness. lets discuss.

**Review by indrajra — CHANGES_REQUESTED** [review]

File: user-org-es-util/src/main/java/org/sunbird/es/service/impl/ElasticSearchRestHighImpl.java (hunk 40 above)
Comment (anmol2302) [reply user-org-es-util/src/main/java/org/sunbird/es/service/impl/ElasticSearchRestHighImpl.java]: done

File: user-org-actor/src/main/java/org/sunbird/Application.java (hunk 22 above)
Comment (anmol2302) [reply user-org-actor/src/main/java/org/sunbird/Application.java:36]: once hari will complete the task.

File: user-org-actor/src/main/java/org/sunbird/BaseActor.java [hunk 56]
```diff
@@ -29,7 +29,7 @@ public void onReceive(Object message) throws Throwable {
             try {
                 ProjectLogger.log(String.format("%s:%s:method started at %s",this.getClass().getSimpleName(),operation,System.currentTimeMillis()), LoggerEnum.DEBUG);
```
Comment (indrajra) [inline user-org-actor/src/main/java/org/sunbird/BaseActor.java:30]: missed use startTrace()?

**Review by indrajra — APPROVED** [review]
